Documentation Update - Accessibility page

yogmel opened the pull request:
## Description

This PR updates the documentation to reflect a11y changes and guidelines, as seen in #5725  and #5781.

Sections added:
- General Practices
  - Reactstrap
  - Font Awesome
- Colors
  - Colors in buttons
  - Inserting new colors
  - Red and Green
- Images
  - Image HTML tag
  - SVG
- Interactive elements
  - Dropdown menu
  - Forms, inputs and buttons
- Specific elements
  - Progress Bar
  - Tables
  - Other Components
- References

yogmel commented:
Just a note: I couldn't test it locally because I was not able to install the right version of `mkdocs` (it is apparently not available for the OS). I used the model to update the Installation and navigation files.

sarah-clements commented on docs/installation.md:
````diff
@@ -40,6 +40,10 @@ production site. You do not need to set up the Docker environment unless making
   yarn start:stage
   ```
 
+### Following code directories
````
Change this to "Code Standards"

sarah-clements replied:
Also, please move this so it's after the "Prerequisites" section.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,58 @@
+|     `text-info` | `text-darker-info`     |
+| `btn-secondary` | `btn-darker-secondary` |
+
+For example, if inserting a text color into an element, use `className="text-darker-secondary"`.
```
Can you add to the end of the sentence " instead of using `prop=secondary`".

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,58 @@
+
+### Inserting new colors
+
+If you are inserting new colors into the Treeherder UI, please make sure the color + background-color combination passes the minimum constrast ratio defined by WCAG.
```
"If you add new colors to the style sheets, please make sure the text and background color combination passes the minimum contrast ratio defined by WCAG."

(also you have a typo with contrast).

sarah-clements commented:
This looks great! Exactly what I was looking for. One thing that would be nice to add (if it isn't already on your list) is a section on when aria labels need to be added and which libraries (like reactstrap and FontAwesomeIcon) have built in aria labels based on props you use (for example, using the `title` prop with FontAwesomeIcon). :)

sarah-clements commented:
We should probably encourage people to use reactstrap as much as possible, instead of standard html elements, so they don't have to remember to add attributes that reactstrap adds "under the hood".

yogmel commented:
> This looks great! Exactly what I was looking for. One thing that would be nice to add (if it isn't already on your list) is a section on when aria labels need to be added and which libraries (like reactstrap and FontAwesomeIcon) have built in aria labels based on props you use (for example, using the `title` prop with FontAwesomeIcon). :)

@sarah-clements thanks! Yes, it is on my list indeed. Do you want me to send these updates on this PR or a following one?

sarah-clements commented:
> > This looks great! Exactly what I was looking for. One thing that would be nice to add (if it isn't already on your list) is a section on when aria labels need to be added and which libraries (like reactstrap and FontAwesomeIcon) have built in aria labels based on props you use (for example, using the `title` prop with FontAwesomeIcon). :)
> 
> @sarah-clements thanks! Yes, it is on my list indeed. Do you want me to send these updates on this PR or a following one?

If you have time to do it, I think it'd be great to see them.

yogmel commented:
> If you have time to do it, I think it'd be great to see them.

Sure, I've pushed the complete documentation update for review.

I really appreciate your help with my word choice.

sarah-clements commented on docs/installation.md:
```diff
@@ -8,6 +8,10 @@
 
 If you only want to hack on the frontend, see the UI Development section below. If you want to hack on the backend or work full-stack, see the [Server and Full-stack Development](#server-and-full-stack-development) section.
 
+### Code Standards
```
This should only have two ##

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+If you can not find a suitable component, prefer [HTML5 semantic tags](https://developer.mozilla.org/en-US/docs/Glossary/Semantics#Semantics_in_HTML) and avoid using generic tags, such as `span` and `div`. Use those only when there is no other more semantic tag. Moreover, try to use more **headings** (`h1` to `h6`) whenever possible, because screen readers users usually navigate by headings and t ...
+
+If you can build your layout neither with `reactstrap` nor with semantic tags, please insert an `aria-role`. Down below, there are some elements constructed that way in Treeherder:
+
```
This wording is a little unclear to me. Since you mentioned that there are elements constructed this way in Treeherder, I was expecting to click on a link to show some Treeherder code to show how it was done. You don't need to do that but can you reword this?

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+As a general rule, accessible software is achieved by following web development best practices. As Treeherder uses the `reactstrap` library, it is recommended using one of its components, if possible.
+
+If you can not find a suitable component, prefer [HTML5 semantic tags](https://developer.mozilla.org/en-US/docs/Glossary/Semantics#Semantics_in_HTML) and avoid using generic tags, such as `span` and `div`. Use those only when there is no other more semantic tag. Moreover, try to use more **headings** (`h1` to `h6`) whenever possible, because screen readers users usually navigate by headings and t ...
+
```
I would omit this sentence " Use those only when there is no other more semantic tag" would be clearer as " and add to the last sentence "because it is not accessible to screen readers" (or however you want to word that).

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+    <Progress bar value={regression ? 100 - magnitude : magnitude} />
+    <Progress bar value={regression ? magnitude : 100 - magnitude} />
+  </div>
+</Progress>
```
Since this is an example, I think it'd be better to simply the code (no ternaries, just add values) so it's easier to follow.

yogmel commented:
@sarah-clements Thanks! It means a lot to me.
I have changed some major parts of the General Practices and added a summary. Please let me know if it reads better.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+### Reactstrap
+
+Some elements in Treeherder's UI are borrowed from the `reactstrap` library. As a general rule, its elements are already accessible.
+That library is built on top of Bootstrap, therefore the [Bootstrap Accessibility documentation](https://getbootstrap.com/docs/4.0/getting-started/accessibility/) is the one to look for, if in doubt.
```
Reactstrap uses bootstrap's CSS but the implementation is purely react/JSX, whereas Bootstrap uses HTML and jquery. So I'm wondering if this is correct - whether we should be directly people to Bootstrap for accessibility concerns.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+Some elements in Treeherder's UI are borrowed from the `reactstrap` library. As a general rule, its elements are already accessible.
+That library is built on top of Bootstrap, therefore the [Bootstrap Accessibility documentation](https://getbootstrap.com/docs/4.0/getting-started/accessibility/) is the one to look for, if in doubt.
+
+There are some issues on **Color Contrast** that are pointed by Bootstrap itself. This was solved in Treeherder by providing [**_custom color classes_**](accessibility.md#colors).
```
I don't think this sentence is needed (and it's not really related to reactstrap per say), so it can be removed.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+
+### Colors in buttons
+
+In [`reactstrap`'s `Button` component](https://reactstrap.github.io/components/buttons/), you can insert a `prop` specifying its color. In that case, you use only the color name, without its prefix.
```
The subtitle isn't necessary - it can be in the "Colors" section.

I think this custom color prop is applicable to any reactstrap component that accepts a color prop, right? Please reword if that's the case. If not, that should be added in a separate pull request.

yogmel replied:
> I think this custom color is applicable to any reactstrap component that accepts a color prop, right? Please reword if that's the case. If not, that should be changed in a separate pull request.

This props can be inserted, but it won't work on all of them. I'll both reword and send a separate PR.

sarah-clements commented on docs/accessibility.md:
````diff
@@ -0,0 +1,221 @@
+If you need to insert an event listener in a non-interactive element, such as a `span`, add also an `aria-role` of `button`, `link`, `checkbox`, or whatever seems closer to the functionality of the element.
+
+```html
+<span onClick="{someFunction}" aria-role="button">
````
Remove the parenthesis around {someFunction}

sarah-clements commented on docs/accessibility.md:
````diff
@@ -0,0 +1,221 @@
+In case the content of the button is not a text (an icon or graphical element, for example), this element should have an `aria-label`, explaining what it does.
+
+```jsx
+<Button aria-label="Bookmark this" onClick={onClickFunction}>
````
Even though this is just an example, we want to encourage people to be very explicit :) So perhaps "bookmark alert 12345" would be better.

sarah-clements commented on docs/accessibility.md:
````diff
@@ -0,0 +1,221 @@
+```
+
+For input and checkbox, those are important to be labeled - placeholders are not enough for screen reader users. For example, when inserting an email input and a checkbox, using `reactstrap`, the structure should be:
+
````
Nit: Can you remove the back ticks for reactstrap? It's not code, just a library name, so it looks odd like this.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+### Progress Bar
+
+Across Perfherder, there are some Progress Bars, which are built using `reactstrap` components. There is still an [issue](https://github.com/reactstrap/reactstrap/issues/1681), however, when adding labels on each bar. While this is still not solved, it is recommended to add an `aria-label` in the wrapping component, explaining the values of the bar.
+
```
I see that you've offered to work on that bug. Well done :) Please make a note to update the documentation when that bug has been fixed. You can file a bugzilla bug if you want.

yogmel replied:
I was going to tell you only when it was solved lol but, well, open source, right?
Ok, will do!

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+
+Also, if you are not using `<table>` tag or `reactstrap`'s `Table` component to create the table, please use [all the necessary `role` attributes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Table_Role).
+
+Alternatively, you may choose to use the `React Table` library. The only requirement here is to insert a `role="table"`, like this:
```
Remove markdown for React Table please and add a link to the docs.

sarah-clements commented on docs/accessibility.md:
````diff
@@ -0,0 +1,221 @@
+```jsx
+<ReactTable
+  getTableProps={() => ({ role: 'table' })}
+  // ...other table's props
````
This comment isn't necessary :)

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+      <th> Third header </th>
+    </tr>
+  </thead>
+  // ...table's body
```
Could add an empty `<body />` instead.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+
+## Colors
+
+Treeherder uses [Bootstrap color](https://getbootstrap.com/docs/4.0/utilities/colors/) utility classes. While most of them work, there are a couple that needed a little tweaking to increase **color contrast**, according to [WCAG standards](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast?utm_source=devtools&utm_medium=a11y-panel-checks-color ...
```
Please swap out the sentence starting with "While most..." with this: "A few were modified in order to meet WCAG standards for color contrast."  It reads a little smoother :)

yogmel replied:
Ok!

> Edit: I noticed the language here implies more than one color was modified, but I think only "info" was right?

There were two, actually:  info and secondary.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+
+Treeherder uses [Bootstrap color](https://getbootstrap.com/docs/4.0/utilities/colors/) utility classes. While most of them work, there are a couple that needed a little tweaking to increase **color contrast**, according to [WCAG standards](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast?utm_source=devtools&utm_medium=a11y-panel-checks-color ...
+
+So, we ask that you substitute the Bootstrap color to the Treeherder one, according to the following table:
```
"Please use this custom color according to the following table:"

sarah-clements replied:
This can be a follow up, but did you make changes to red and green or was it only for specific Treeherder buttons on the Job View UI (I can't remember)?

yogmel replied:
>    This can be a follow up, but did you make changes to red and green or was it only for specific Treeherder buttons on the Job View UI (I can't remember)?

I did change their colors, but the reds and greens and the other colors on the Jobs View were not from Bootstrap, they were custom made. They are in `treeherder-job-buttons.css`.

sarah-clements commented on docs/accessibility.md:
```diff
@@ -0,0 +1,221 @@
+
+## General Practices
+
+As a general rule, accessible software is achieved by following web development best practices. As Treeherder uses the `reactstrap` library, it is recommended using one of its components, if possible.
```
Also add a link to the reactstrap docs.

yogmel commented:
Thank you all :)

@sarah-clements I will send a PR about the color props shortly, so it's not a blocker.
